### PR TITLE
Implement `knownForType` for RFC#58.

### PR DESCRIPTION
**Pull request by rwjblue**
Supports https://github.com/emberjs/ember.js/pull/11393.

Further reading: https://github.com/emberjs/rfcs/pull/58.

### CODE REVIEW

**Comment by rwjblue** [comment]
I split this into two commits to make it easier to review.  The first commit is the actual change, and the second is updating `dist/`.

File: dist/ember-resolver.js [hunk 1]
```diff
@@ -337,6 +336,45 @@ define("ember/resolver",
       }
 
       Ember.Logger.info(symbol, parsedName.fullName, padding, description);
+    },
+
+    eachForType: function(type, callback) {
+      var moduleEntries = requirejs.entries;
+
+      for (var moduleName in moduleEntries) {
```
Comment (stefanpenner) [inline dist/ember-resolver.js]: `Ember.keys`

**Comment by stefanpenner** [comment]
i believe in the past the idea was `Array.isArray(resolveAllForType(name)) === true`

**Comment by rwjblue** [comment]
@stefanpenner - I don't fully understand your suggestion. Can you elaborate? What would 'resolveAllForType' return (or do)?

**Comment by rwjblue** [comment]
The current primary use-case for this is to enumerate the list of names for the current type for the dashless helper RFC. In that case we actually do not have a use for the factories (since we are just whitelisting the helper names themselves), but I added them here because I think we will find that useful in the future. 

I could definitely see this method returning an object instead of calling a callback if you would prefer that, but I don't see how an array would work (since we need both a key and a value).

**Comment by stefanpenner** [comment]
> @stefanpenner - I don't fully understand your suggestion. Can you elaborate? What would 'resolveAllForType' return (or do)?

* So this feature looks useful for more then just dashless helpers, but also initializers and instance initializers.

* Also, rather then locking us into having to use `forEach` to enumerate, by returning an array (instead of just exposing forEach), We can also filter/map/inject/for in/for on the collection. Basically it just feels more flexible.

**Comment by rwjblue** [comment]
@stefanpenner - To confirm, you would prefer something like the following:

```
var items = resolver.eachForType('fruit');

/*
  Where items is something like:
  {
    'fruit:apple': AppleFactory,
    'fruit:orange': OrangeFactory
  }
*/
```

Right?

**Comment by teddyzeenny** [comment]
Looks similar to what [container-debug-adapter](https://github.com/ember-cli/ember-resolver/blob/3eb838155f9b3165b41895bdb51f510520a10d3c/packages/ember-resolver/lib/container-debug-adapter.js) does. Should we consider removing the container debug adapter?

**Comment by rwjblue** [comment]
@teddyzeenny - I think that could be refactored to use the new method here, but I'm not sure that it completely replaces the container debug adapter.

**Comment by stefanpenner** [comment]
> Right?

i was actually just thinking an array of names, but an array of well formed pojos is also really cool

**Comment by rwjblue** [comment]
@stefanpenner - Updated to address your concerns (as I understood them):

* Swapped to using `Ember.keys` instead of `for`/`in`.
* Remove callback argument so a plain dictionary is returned.

Can you double check, and :+1: / :-1:?

**Comment by stefanpenner** [comment]
> Remove callback argument so a plain dictionary is returned.

as mentioned above, an array of well formed pojos seems like the must flexible return value.

**Comment by rwjblue** [comment]
@stefanpenner - Not sure that I agree.

How is an array of key/value objects better than an actual structure that is built to deal with keys and values (a single object)?

```javascript
[
  {key: 'fruit:apple', value: AppleFactory},
  {key: 'fruit:orange', value: OrangeFactory}
]
```

vs

```javascript
{
  'fruit:apple': AppleFactory,
  'fruit:orange': OrangeFactory
}
```

Are you concerned with needing to add more things than key (container fullname) and value (actual factory)?

**Comment by rwjblue** [comment]
In the end I just want to settle this so we can move forward on the dashless helpers RFC which needs to be out in 1.13 tomorrow.

If you are certain that you want an array of objects, I will change to that API.

**Comment by stefanpenner** [comment]
> How is an array of key/value objects better than an actual structure that is built to deal with keys and values (a single object)?

the common pattern is to enumerate/filter/map/transform this information. It seems like exposing an iterable primitive is appropriate.

**Comment by stefanpenner** [comment]
> helpers RFC which needs to be out in 1.13 tomorrow.

this is true, but once we land this it will be around for ages, so being mindful and diligent now saves us from prolonged pain

**Comment by rwjblue** [comment]
> this is true, but once we land this it will be around for ages, so being mindful and diligent now saves us from prolonged pain

Please remember that this is also a private API.

**Comment by rwjblue** [comment]
Updated to use an array of objects as requested.

**Comment by stefanpenner** [comment]
hmm, I just realized these iterables may yield duplicates and we have lost the precedence provided  during resolving.

for example:

```
app/components/edit-user.js
app/pods/edit-user/componeont.js
```

clearly in resolving, we realize the later shadows the former.

**Comment by rwjblue** [comment]
Yes, using the object form automatically de-dupes for us and allows us to ensure the proper ordering.

If your primary concern with the object case is future expansion in what the values should be (for example maybe we also want `modulePath`), I could make something like:

```javascript
{
  'fruit:apple': {key: 'fruit:apple', value: AppleFactory},
  'fruit:orange': {key: 'fruit:orange', value: OrangeFactory}
}
```

That still uses an object for de-duping and prioritization, but allows expansion on the number of keys in the object later...

**Comment by stefanpenner** [comment]
@rwjblue the object form clearly dedupes, but I am unclear where in the code it takes the priority into account

**Comment by rwjblue** [comment]
This PR only addresses the resolver parts, not registry or fallback registry.  That is done in the Ember PR which will have the order (later entries overriding earlier ones) as fallback registry, registry,  resolver.

I do not currently handle ordering of different module types in this PR. So if you had both a pod style and non-pod style module with the same name whichever was last in `requirejs.entries` would win out. I could add a queueing system that works like `moduleNameLookupPatterns` [here](https://github.com/ember-cli/ember-resolver/blob/3eb838155f9b3165b41895bdb51f510520a10d3c/packages/ember-resolver/lib/core.js#L228-L235) if you would like.

**Comment by stefanpenner** [comment]
> I could add a queueing system that works like moduleNameLookupPatterns here if you would like.

I would be willing to forgo this if you can tell me why its not important. It seems important?

**Comment by rwjblue** [comment]
> It seems important?

It is not important for the current need.  We currently only need to know the names of the types in the system, and we do absolutely nothing with the factory or other information returned.  All we need to know is whether a specific name exists (nothing more).  

My needs for RFC#58 would be satisfied if `eachForType` simply returned a POJO like: 

```javascript
{
  'helper:t': true,
  'helper:blah': true
}
```

---

We *could* go with a simple true value (it is the minimum requirement ATM), we can always "upgrade" it to an object if we need to, and any usage of it (like `if (returnValueFromEachForType['helper:t']) { }`) would still work if/when we need to flesh it out as an object with more values (since that would also be truthy)...

**Comment by stefanpenner** [comment]
> It is not important for the current need. We currently only need to know the names of the types in the system, and we do absolutely nothing with the factory or other information returned. All we need to know is whether a specific name exists (nothing more).

Ah this was the missing piece of information. Thank you!

**Comment by stefanpenner** [comment]
this LVGTM, :+1:

**Comment by rwjblue** [comment]
Renamed to `knownForType`, and went with simple `true` value dictionary.

**Comment by stefanpenner** [comment]
sg

**Comment by mixonic** [comment]
lgtm :-D

**Comment by stefanpenner** [comment]
thanks for going on this journey with me

File: dist/ember-resolver.js [hunk 2]
```diff
@@ -115,23 +115,22 @@ define("ember/resolver",
...
+      var defaultExport = this._extractDefaultExport(normalizedModuleName, parsedName);
 
-      if (module && module['default']) { module = module['default']; }
-
-      if (module === undefined) {
+      if (defaultExport === undefined) {
         throw new Error(" Expected to find: '" + parsedName.fullName + "' within '" + normalizedModuleName + "' but got 'undefined'. Did you forget to `export default` within '" + normalizedModuleName + "'?");
       }
 
       if (this.shouldWrapInClassFactory(module, parsedName)) {
```
Comment (teddyzeenny) [inline dist/ember-resolver.js]: should `module` be rename to `defaultExport` here?

Comment (rwjblue) [inline dist/ember-resolver.js]: Only if we would prefer to avoid an error.

**Pull request merged**